### ISSUE TITLE
Unclear messaging from pinpoint tryjobs and hard to find results

### ISSUE THREAD

**Issue by perezju** [catapult-project/catapult#4298]
*Writing down some of my thoughts while going through a common "user journey", hoping to spark some discussion on things that could be improved.*

On [issue 816524](https://bugs.chromium.org/p/chromium/issues/detail?id=816524) a pinpoint try job was launched to assess whether the fix for a performance regression would work.

When the job finished it posted to the bug:

> 📍 Couldn't reproduce a difference.
> https://pinpoint-dot-chromeperf.appspot.com/job/16a0a288440000 

which doesn't sound very reassuring.

Visiting the url shows a huge graph with two dots both at a value of 1.0, unclear what the y-axis means, and a not reassuring but prominent **No differences found**.

After some clicking around I figured that "couldn't reproduce a difference" actually meant it couldn't reproduce a *functional* difference, and 1.0 is probably the "success" result. Which in this case is a good start: the build wasn't broken before, and it's still now broken now, good!

It took me little longer to figure out that there was a small [Analyze benchmark results](https://pinpoint-dot-chromeperf.appspot.com/results2/16a0a288440000) link near the top, which finally showed me some of the information I was looking for: without the patch two CPUs were being woken up, with the patch only one of them is. Probably a **huge and very relevant difference**, precisely what I was looking for.

**Comment by simonhatch** [comment]
I wonder if we should just go ahead and show the results2 automatically? Maybe iframe it below or something?

**Comment by dave-2** [comment]
Yes, I was just thinking that the problem is that the perf tryjob ui simply doesn't exist. There's only the bisect ui.

Putting the results2 where the chart is now is probably the best thing we can do right now. We need https://chromium-review.googlesource.com/c/catapult/+/935146 to achieve acceptable latency and reliability for it.

We can also update the bug comment differently for perf tryjobs and bisect jobs.

For the "try a patch with this configuration" button on the Job page, I'm thinking of recategorizing those as bisect jobs, since the behavior and UI should be closer to a bisect job than a try jobs.

**Comment by perezju** [comment]
Very good ideas, they sound good to me.

Wondering on the "try a patch with this configuration" on the Job page, what does it exactly do?

**Comment by dave-2** [comment]
The use cases we're trying to cover are:
* Functional and performance bisects.
* Try jobs where the user wants to compare all the results of a benchmark (results2) for a given patch.
* Try jobs where the user wants to compare all the results of a benchmark between two milestones.
* "Try a patch with this configuration" where the user's CL was named as a culprit and they want to see whether a patch they're working on fixes the regression. I think this is more like a bisect job since we have a bisection metric and want the repeat count to increase (see below).

So to be clear, the differences are:

|| Try job | Bisect job
--- | --- | ---
**`auto_explore`** | `False` | `True`
**Metric** | Should not specify a bisection metric. | Must specify a bisection metric (functional or performance).
**Repeats** | Fixed at 10 repeats. | Increases the number of repeats until we decide that metric does/doesn't have a difference.
**Changes** | The list of Changes doesn't change. | Automatically adds Changes to narrow down the regression. May also add Changes outside the range. (Doesn't do anything if we're comparing with/without patch since the base commits are the same.)
**Job page** | Shows results2. | Shows the results chart. #3877
**Bugs** | Posts "Try job complete. See results below." on the bug and doesn't auto-assign. | Posts "differences found/not found" on the bug and auto-assigns.

I'm also thinking maybe replace `auto_explore`, with a trinary `bisection_mode` field that can be `"functional"`, `"performance"`, or `None`. `None` indicates a try job. Then we can use the same field to distinguish between try and bisect, and between functional and performance for #4283.

**Comment by anniesullie** [comment]
Thanks for putting this together! I think it's mostly right, except:

* One more use case: getting a trace with more debug categories

So I think the things that aren't quite right are:
* Sometimes the user really does want a metric. Cases I can think of are:
  * They've been asked to run a try job before committing a new benchmark and want the result code as a metric.
  * They're trying to re-enable a benchmark which was disabled for failing, and they want to run a perf tryjob to see if it succeeds, so they want the result code as a metric.
  * They're trying to reproduce a perf regression bug assigned to them.
  * They're trying to get a debug trace on a perf regression bug assigned to them, but before they look at the trace the first thing they want to see is whether it reproduced the regression.
* Sometimes perf try job users want more repeats. I know I've seen pasko ask for this.
* Sometimes the user is trying to get directly to a trace ("request debug trace" button) and showing them a results2 makes that a bit harder.

So I think perf tryjobs are really hard because the users actually have three different things in mind:
* Some users were expecting to see a specific metric. I think we can mostly catch these ahead of time (buttons on perf dashboard/pinpoint, have a "functional" option in the dialog and have benchmark documentation link to a dialog with it set). I am not sure if we want to use the same UI as we do for bisection when we show the bisection result.
* Some users were expecting to see a trace. But they really do want to also see right away if the issue reproduced as expected. We can identify these users because they used "request debug trace" button.
* Some users were expecting a results2. I think these are the users that showed up at the dialog without coming through any specific workflow on dashboard/pinpoint/benchmarking documentation.

**Comment by dave-2** [comment]
> Sometimes the user really does want a metric.

Sorry, I was unclear. I meant that we would categorize all of these as "bisect" jobs. The name is a misnomer because it doesn't perform a bisection (doesn't add additional Changes) if you run it with HEAD and HEAD+patch. We can call it something else. I also removed "auto-assign" from the behavior differences; in both cases, the user may want to specify whether to auto-assign.

> Sometimes perf try job users want more repeats. I know I've seen pasko ask for this.

I've updated the chart. The difference isn't the number of repeats, but the behavior. Try jobs have a manual (user-specified) repeat count and bisect jobs have an automatic (determined by comparing the bisection metric) repeat count.

> Sometimes the user is trying to get directly to a trace ("request debug trace" button) and showing them a results2 makes that a bit harder.

The tracing UI is awkward in both cases, and we may have to think of something better. Since the results2 will be iframed into the Job page we have flexibility on what shows up around it.

**Comment by perezju** [comment]
Also thanks for this! The table above looks like a very good summary to make things clearer.

Agree with Annie's observation that there are at least three use cases for try jobs. But maybe all three can live within the same UI, as long as there are clear and prominent labeled-tabs/buttons to access them? E.g. "Functional results", "Benchmark measurements" (results.html) and "View individual traces". (Or something like that, I'm not a UI person :stuck_out_tongue:)

> I'm also thinking we replace auto_explore with a trinary bisection_mode field that can be "functional", "performance", or None. None indicates a try job. Then we can use the same field to distinguish between try and bisect, and between functional and performance for #4283.

Nit: I think more explicit is better, maybe there should also be a `job_type` that can be either `'bisect'` or `'try_job'`?

**Comment by dave-2** [comment]
What I want to codify is that sometimes the user cares about a comparison metric, and sometimes there isn't one. And all the differences in the table follow from that. So a `comparison_mode` field would make that explicit. Right now the tryjob/bisect job dichotomy doesn't fit quite right, since some things we currently consider "tryjobs" do have a comparison metric and should act more like "bisect jobs".

**Comment by anniesullie** [comment]
Agreed with the above, sounds great!
